tracing-opentelemetry/src/layer.rs

```diff
@@ -896,6 +942,52 @@ where
                 } else {
                     builder.events = Some(vec![otel_event]);
                 }
+
+                // limit the maximum number of events we'll record per-span
+                match (self.max_events_per_span, &mut builder.events) {
+                    (Some(max_num_events), Some(ref mut events))
+                        if events.len() > max_num_events =>
+                    {
+                        // drain half of the existing events, using ceiling division
+                        // to handle the case where max_num_events is 0 or 1
+                        //
+                        // note: this should be more efficient than just continuously
+                        // popping the first element
+                        let num_events_to_drain = (events.len() + 2 - 1) / 2;
+                        events.drain(0..num_events_to_drain);
+
+                        Some(num_events_to_drain)
+                    }
+                    _ => None,
+                }
+            } else {
+                None
+            };
+
+            // record how many events we dropped
+            match (dropped_events, extensions.get_mut::<DroppedOtelEvents>()) {
+                (Some(additional_dropped), Some(current_dropped)) => {
+                    current_dropped.add_dropped(additional_dropped);
+
+                    // report, once per span, when we drop events
+                    if !current_dropped.reported {
+                        let id = span.id().into_u64();
+                        let name = span.name();
+                        let module = span.metadata().module_path();
+                        let line = span.metadata().line();
+
+                        tracing::error!(
+                            "Dropped events for span({id}) named: {name} at {module:?}:{line:?}"
+                        );
```

Comment:
a tracing event inside a tracing layer is a recipe for disaster i think...

In the context of mz, an `eprintln!` might work, but getting that merged upstream is impossible...can we instead fabricate one extra otel event and attach it to the otel span `on_close`?

Reply:
Yeah I was definitely weary of this 😅

I changed the impl to add an attribute during `on_close`, that's what we seemed to do with `Timings`, what do you think about that?

Reply:
sounds perfect!